src/wp-includes/media.php

```diff
@@ -5011,9 +5012,132 @@ function wp_getimagesize( $filename, array &$image_info = null ) {
+ * @return array $webp_info {
+ *     An array of WebP image information.
+ *
+ *     @type array $size {
+ *         @type int  $width  Image width.
+ *         @type int  $height Image height.
+ *         @type bool $type   The WebP type: one of 'lossy', 'lossless' or 'animated-alpha'.
+ *     }
+
+ */
+function wp_get_webp_info( $filename ) {
+try {
+	$handle = fopen( $filename, 'rb' );
+	if ( $handle ) {
+		$magic = fread( $handle, 40 );
+		fclose( $handle );
+
+		// Make sure we got enough bytes.
+		if ( strlen( $magic ) < 40 ) {
+			return false;
+		}
+
+		$width  = false;
+		$height = false;
+		$type   = false;
+
+		// The headers are a little different for each of the three formats.
+		switch ( substr( $magic, 12, 4 ) ) {
+			// Lossy WebP.
+			case 'VP8 ':
+				$parts  = unpack( 'v2', substr( $magic, 26, 4 ) );
+				$width  = (int) ( $parts[1] & 0x3FFF );
+				$height = (int) ( $parts[2] & 0x3FFF );
+				$type   = 'lossy';
+				break;
+			// Lossless WebP.
+			case 'VP8L':
+				$parts  = unpack( 'C4', substr( $magic, 21, 4 ) );
+				$width  = (int) ( $parts[1] | ( ( $parts[2] & 0x3F ) << 8 ) ) + 1;
+				$height = (int) ( ( ( $parts[2] & 0xC0 ) >> 6 ) | ( $parts[3] << 2 ) | ( ( $parts[4] & 0x03 ) << 10 ) ) + 1;
+				$type   = 'lossless';
+				break;
+			// Animated/alpha WebP.
+			case 'VP8X':
+				// Pad 24-bit int.
+				$width = unpack( 'V', substr( $magic, 24, 3 ) . "\x00" );
+				$width = (int) ( $width[1] & 0xFFFFFF ) + 1;
+				// Pad 24-bit int.
+				$height = unpack( 'V', substr( $magic, 27, 3 ) . "\x00" );
+				$height = (int) ( $height[1] & 0xFFFFFF ) + 1;
+				$type   = 'animated-alpha';
+				break;
+			if ( $width && $height && $type ) {
+				return array(
+					'width'  => $width,
+					'height' => $height,
+					'type'   => $type,
+				);
+			}
+		}
+	} catch ( Exception $e ) {
+	}
+	return array();
+}
+
+/**
+ * Determine if a passed image is a lossy WebP image.
+ *
+ * @param [type] $filename
+ * @return void
+ */
+function _wp_webp_is_lossy( $filename ) {
+	$webp_info = wp_get_webp_info( $filename );
+	$type      = isset( $webp_info['type'] ) ? $webp_info['type'] : false;
+	return $type && 'lossy' === $type;
+}
+
+/**
+ * Get the image size, with support for WebP images.
+ *
+ * @since 5.8.0
+ * @access private
+ *
+ * @param string $filename  The file path.
+ * @param array  $imageinfo Extended image information, passed by reference.
+ */
+function _wp_get_image_size( $filename, &$imageinfo = array() ) {
+	// Try getimagesize() first.
+	$info = getimagesize( $filename, $imageinfo );
+	if ( false !== $info ) {
+		return $info;
+	}
+	// For PHP versions that don't support WebP images, extract the image
+	// size info from the file headers.
+	if ( 'image/webp' === wp_get_image_mime( $filename ) ) {
+		$webp_info = wp_get_webp_info( $filename );
+		$width     = isset( $webp_info['width'] ) ? $webp_info['width'] : false;
+		$height    = isset( $webp_info['height'] ) ? $webp_info['height'] : false;
```

Comment:
https://github.com/WordPress/wordpress-develop/pull/1137/files#r609091846

Reply:
not sure what this comment refers to?

Reply:
`wp_getimagesize()` was added in 5.7.0 https://github.com/WordPress/wordpress-develop/blob/8603c6a00615952563ed35e101f2c65b424c6f8a/src/wp-includes/media.php#L4988 . May want to move the WebP code into it.

Reply:
Again, `wp_get_webp_info` always returns with width and height set now. Why is this needed?

Reply:
ah right @spacedmonkey - the check is no longer needed, will remove.

Reply:
@hellofromtonya - right, thanks! This new code integrates into that function adding support for WebP even if the underlying server does not have WebP support.

Reply:
Removed `isset` checks.